### PR TITLE
Add GroupsApi test

### PR DESCRIPTION
**Pull request by toddbellMSFT**
Recent investigation required me to write this test, thought we may want to add it to our test suite if it is logical/correct to just create groups.

### CODE REVIEW

File: source/test/TestApp/PlayFabApiTest.cpp [hunk 1]
```diff
@@ -16,11 +16,48 @@
...
 using namespace AuthenticationModels;
 using namespace DataModels;
 
+namespace uuid {
+    static std::random_device              rd;
+    static std::mt19937                    gen(rd());
+    static std::uniform_int_distribution<> dis(0, 15);
+    static std::uniform_int_distribution<> dis2(8, 11);
+
+    std::string generate_uuid_v4() {
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabApiTest.cpp:33]: Generating a random guid would be a great general utility, not just a random method in a test.

File: source/test/TestApp/PlayFabApiTest.cpp [hunk 2]
```diff
@@ -730,6 +767,43 @@ namespace PlayFabUnit
         }
     }
 
+     // GROUPS API
+    void PlayFabApiTest::GroupsApiTest(TestContext& testContext)
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabApiTest.cpp:771]: This new set of tests should be a whole new test file: PlayFabGroupsTest.cpp/.h

I would have many more comments and changes to suggest if we don't do that and leave it here. But I think if we separate this to a separate file & test suite, we'd negate those other comments, so I'll just make this suggestion alone.

File: source/test/TestApp/PlayFabApiTest.cpp [hunk 3]
```diff
@@ -730,6 +767,43 @@ namespace PlayFabUnit
...
+    {
+        auto settings = std::make_shared<PlayFabApiSettings>();
+        //clientApi = std::make_shared<PlayFab::PlayFabClientInstanceAPI>(settings);
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabApiTest.cpp:774]: Delete unused code comments.
Also, this is definitely something that should only exist in a setup step, not in the middle of a test.

**Review by pgilmorepf — CHANGES_REQUESTED** [review]
Main comment is just to create a new suite and put the new tests there.

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    void PlayFabGroupsTest::AddTests()
+    {
+        AddTest("CreateGroupApi", &PlayFabGroupsTest::GroupsApiTest);
+    }
+
+    void PlayFabGroupsTest::ClassSetUp()
+    {
+        clientApi = std::make_shared<PlayFabClientInstanceAPI>(PlayFabSettings::staticPlayer);
+        PlayFabSettings::staticSettings->titleId = testTitleData.titleId;
+        USER_EMAIL = testTitleData.userEmail;
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:58]: This test does not need USER_EMAIL

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        // Verify all the inputs won't cause crashes in the tests
+        TITLE_INFO_SET = !PlayFabSettings::staticSettings->titleId.empty() && !USER_EMAIL.empty();
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:61]: This test does not need USER_EMAIL

File: source/test/TestApp/PlayFabGroupsTest.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#include "playfab/PlayFabGroupsDataModels.h"
+
+namespace PlayFabUnit
+{
+    class PlayFabGroupsTest : public PlayFabApiTestCase
+    {
+    private:
+        // Fixed values provided from testInputs
+        bool TITLE_INFO_SET;
+        std::string USER_EMAIL;
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.h:19]: This test does not need USER_EMAIL

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    void PlayFabGroupsTest::GroupsTestGroupCallback(const GroupsModels::CreateGroupResponse&, void* customData)
+    {
+        TestContext* testContext = static_cast<TestContext*>(customData);
+        testContext->Pass();
+    }
+
+    void PlayFabGroupsTest::GroupsTestLoginFailedCallback(const PlayFabError& error, void* customData)
+    {
+        TestContext* testContext = static_cast<TestContext*>(customData);
+        testContext->Fail("Expected Group Login to succeed. Got error " + error.ErrorMessage);
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:46]: error.GenerateErrorReport()

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        groupApi->CreateGroup(req, Callback(&PlayFabGroupsTest::GroupsTestGroupCallback), Callback(&PlayFabGroupsTest::GroupsTestLoginFailedCallback), customData);
+    }
+
+    void PlayFabGroupsTest::GroupsTestGroupCallback(const GroupsModels::CreateGroupResponse&, void* customData)
+    {
+        TestContext* testContext = static_cast<TestContext*>(customData);
+        testContext->Pass();
+    }
+
+    void PlayFabGroupsTest::GroupsTestLoginFailedCallback(const PlayFabError& error, void* customData)
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:43]: Rename:
If this is meant to be a shared callback (as it is used), then it should be renamed to be described as a shared callback, and the error message 3 lines down should be restated to not indicate an explicit login failure.
If this is NOT meant to be a shared callback (as it is named & strings within indicate), then you need to write a separate method as the fail callback for the CreateGroup call.

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        auto groupApi = std::make_shared<PlayFabGroupsInstanceAPI>(result.authenticationContext);
+        auto req = GroupsModels::CreateGroupRequest();
+        req.CustomTags = std::map<std::string, std::string>();
+
+        // TODO Bug 29786037: this map is required to be filled to not get a 500 error with the CreateGroup api call
+        req.CustomTags.insert(std::pair<std::string, std::string>("One", "Two"));
+
+        req.GroupName = GenerateUuidV4();
+        groupApi->CreateGroup(req, Callback(&PlayFabGroupsTest::GroupsTestGroupCallback), Callback(&PlayFabGroupsTest::GroupsTestLoginFailedCallback), customData);
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:34]: This reuses the GroupsTestLoginFailedCallback callback, which by name, and by text within that method doesn't reflect it as generic reusable. See the mirror comment below on that method.

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    void PlayFabGroupsTest::GroupsTestLoginFailedCallback(const PlayFabError& error, void* customData)
+    {
+        TestContext* testContext = static_cast<TestContext*>(customData);
+        testContext->Fail("Expected Group Login to succeed. Got error " + error.ErrorMessage);
+    }
+
+    void PlayFabGroupsTest::AddTests()
+    {
+        AddTest("CreateGroupApi", &PlayFabGroupsTest::GroupsApiTest);
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:51]: I think a better plan here would be to separate the login step into it's own test, and put two tests here.
That gives each call an independent test-timeout, and is a little safer in terms of validating each call independently.
It's true that the "login" test isn't a test of groups, but as we add tests, we don't want to duplicate login logic.

For individual suites we've been very comfortable with the in-suite ordering being strict and reliable.

File: source/test/TestApp/TestUtils.cpp [hunk 11]
```diff
@@ -3,8 +3,44 @@
 #include "TestAppPch.h"
 #include "TestUtils.h"
 
+#include <random>
+#include <sstream>
+
 namespace PlayFabUnit
 {
+    
+    std::string GenerateUuidV4() {
```
Comment (pgilmorepf) [inline source/test/TestApp/TestUtils.cpp:12]: I think that "TestUtils" is a big improvement, but I'll note that the main SDK and thus most customers still won't have direct access to this.
I think I would have been happier with a generic PlayFabUtil.GenerateUuidV4.

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    void PlayFabGroupsTest::GroupsTestLoginCallback(const ClientModels::LoginResult& result, void* customData)
+    {
+        auto groupApi = std::make_shared<PlayFabGroupsInstanceAPI>(result.authenticationContext);
+        auto req = GroupsModels::CreateGroupRequest();
+        req.CustomTags = std::map<std::string, std::string>();
+
+        // TODO Bug 29786037: this map is required to be filled to not get a 500 error with the CreateGroup api call
+        req.CustomTags.insert(std::pair<std::string, std::string>("One", "Two"));
+
+        req.GroupName = GenerateUuidV4();
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:33]: Because this is the only usage of GenerateUuidV4, and as I state down there, it's not "sufficiently random" to be a GUID, I think we should re-evaluate either the name of this method, or its implementation.

File: source/test/TestApp/TestUtils.cpp [hunk 13]
```diff
@@ -3,8 +3,44 @@
...
+        std::random_device              rd;
+        std::mt19937                    gen(rd());
```
Comment (pgilmorepf) [inline source/test/TestApp/TestUtils.cpp:14]: std::mt19937 defines a "source" with 32 bits of randomness.
https://en.cppreference.com/w/cpp/numeric/random/mersenne_twister_engine

File: source/test/TestApp/TestUtils.cpp [hunk 14]
```diff
@@ -3,8 +3,44 @@
...
+        std::uniform_int_distribution<> dis(0, 15);
+        std::uniform_int_distribution<> dis2(8, 11);
+
+        std::stringstream ss;
+        int i;
+        ss << std::hex;
+        for (i = 0; i < 8; i++) {
+            ss << dis(gen);
```
Comment (pgilmorepf) [inline source/test/TestApp/TestUtils.cpp:22]: By calling the same randomizer 4 times in sequence, you're not actually generating guid's as randomly as guids are supposed to be.
Guids have 128 bits of randomness, but in your case, your method only has 32.
Repeatedly calling dis() sequentially won't be 128 bits of randomness, but only 32, because only the first number is "random", since it was seeded by the time. Every number after the first one is predicatable, and based on only the first number.

Meaning that this GUID generator only has 4E9 (4 billion) possible outputs rather than the intended 3E38 (300 undecillion or 300 billion billion billion billion) outputs.

In short, it generates a number that is in guid format, but it's not actually a reliable guid in terms of being super unique. In fact, it's extremely likely to be not unique at all, and if our customers rely on these values being unique, they'll very quickly be disappointed.

**Review by pgilmorepf — CHANGES_REQUESTED** [review]
I think that overall the intent here is very good, but the GUID generator fails to meet the implied expectations.
EITHER: We make a real GenerateUuidV4 method, which is truly unique to 128 bits, or we rename/move the existing method to simply say "GenerateRandomGroupName".

This implementation is more than fine for a random test group name, but insufficient for true GUIDs, if we want to make them visible to customers.

I apologize that this is contradictory to my earlier instructions on the GUID method, but I didn't look at the implementation closely enough last time.

File: source/test/TestApp/TestUtils.cpp (hunk 13 above)
Comment (pgilmorepf) [reply source/test/TestApp/TestUtils.cpp:14]: The simplest improvement is to convert to the 64 bit mt19937_64 option.
Upgrading that to 128 bits is awkward though, as it requires creating two of these, in a way where one isn't biased against the other. I think that it requires putting a sleep between the two, such as a 1ms sleep (which will be randomized by how long the operating system actually chooses to sleep, likely >1ms). That MAY be "sufficient" to get "more than" 64 bits of randomness, but I'm not sure it'll be the full 128 bits. You might have to do some research on how the seed is generated in order to get randomness at the full promised 128 bits.

Comment (toddbellMSFT) [reply source/test/TestApp/TestUtils.cpp:15]: I pulled this info from stack overflow. I'm leaving a note about the source and moving on. This is sufficient for the test we are adding.

File: source/test/TestApp/TestUtils.cpp (hunk 14 above)
Comment (toddbellMSFT) [reply source/test/TestApp/TestUtils.cpp:23]: renamed to not call it a guid.

File: source/test/TestApp/PlayFabGroupsTest.cpp (hunk 10 above)
Comment (toddbellMSFT) [reply source/test/TestApp/PlayFabGroupsTest.cpp:51]: Other login tests in previous test suites have been on a per test basis. 

If we want to combine logins for additional tests in the future we can do it then.

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+using namespace PlayFab;
+
+namespace PlayFabUnit
+{
+
+    // GROUPS API
+    void PlayFabGroupsTest::GroupsApiTest(TestContext& testContext)
+    {
+        auto settings = std::make_shared<PlayFabApiSettings>();
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:16]: Unused.

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    void PlayFabGroupsTest::AddTests()
+    {
+        AddTest("CreateGroupApi", &PlayFabGroupsTest::GroupsApiTest);
+    }
+
+    void PlayFabGroupsTest::ClassSetUp()
+    {
+        clientApi = std::make_shared<PlayFabClientInstanceAPI>(PlayFabSettings::staticPlayer);
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:56]: Looks like it may have been intended to use the static settings from the test above, but failed to move it here.

Comment (pgilmorepf) [reply source/test/TestApp/PlayFabGroupsTest.cpp:56]: Likewise, it's better test isolation if this is initialized with a local settings and context, rather than using staticPlayer and staticContext, which can have interference from other tests.

**Review by pgilmorepf — CHANGES_REQUESTED** [review]
One more small adjustment.
Let's finish the test isolation by using local settings and context, rather than staticSettings and staticPlayer.

File: source/code/source/playfab/QoS/PlayFabQoSApi.cpp [hunk 17]
```diff
@@ -138,7 +138,7 @@ namespace PlayFab
             bool allTimedOut = true;
             for(auto& regionRes: result.regionResults)
             {
-                if(regionRes.latencyMs != timeoutMs)
+                if(regionRes.latencyMs != static_cast<int>(timeoutMs))
                 {
```
Comment (MichaelMackendMS) [inline source/code/source/playfab/QoS/PlayFabQoSApi.cpp:142]: what is this block of code checking exactly?  is there some special value for latencyMs that timeoutMs needs to match in order to be considered timed out?

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        req.CustomTags = std::map<std::string, std::string>();
+
+        // TODO Bug 29786037: this map is required to be filled to not get a 500 error with the CreateGroup api call
+        req.CustomTags.insert(std::pair<std::string, std::string>("One", "Two"));
+
+        req.GroupName = GenerateRandomString();
+        groupApi->CreateGroup(req, Callback(&PlayFabGroupsTest::GroupsTestGroupCallback), Callback(&PlayFabGroupsTest::GroupsTestSharedFailureCallback), customData);
+    }
+
+    void PlayFabGroupsTest::GroupsTestGroupCallback(const GroupsModels::CreateGroupResponse&, void* customData)
```
Comment (MichaelMackendMS) [inline source/test/TestApp/PlayFabGroupsTest.cpp:35]: This would appear to be strictly a success callback.  Can we rename this to more accurately reflect it's purpose?

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    void PlayFabGroupsTest::ClassSetUp()
+    {
+        groupsTestSettings = std::make_shared<PlayFabApiSettings>();
+        groupsTestSettings->titleId = testTitleData.titleId;
+
+        clientApi = std::make_shared<PlayFabClientInstanceAPI>(groupsTestSettings);
+
+        // Bug 45669 - this is required for groups tests to pass.
```
Comment (MichaelMackendMS) [inline source/test/TestApp/PlayFabGroupsTest.cpp:59]: Why is this a bug?

File: source/test/TestApp/TestUtils.cpp [hunk 20]
```diff
@@ -3,8 +3,45 @@
 #include "TestAppPch.h"
 #include "TestUtils.h"
 
+#include <random>
```
Comment (MichaelMackendMS) [inline source/test/TestApp/TestUtils.cpp:6]: Seems like this is also using a pch.  Any reason not to put these includes into the pch?

**Review by MichaelMackendMS — CHANGES_REQUESTED** [review]

File: source/code/source/playfab/QoS/PlayFabQoSApi.cpp (hunk 17 above)
Comment (toddbellMSFT) [reply source/code/source/playfab/QoS/PlayFabQoSApi.cpp:142]: If we time out, regionRes.latencyMs is expected to have hit the timeout that users have set elsewhere (Passed in as a parameter). Rather than change the signature elsewhere (which I think may be a public facing API), I silenced a warning by adding this static cast.

File: source/test/TestApp/PlayFabGroupsTest.cpp (hunk 19 above)
Comment (toddbellMSFT) [reply source/test/TestApp/PlayFabGroupsTest.cpp:59]: without that map, there is an issue on the server that returns a 500 error (not an "invalid parameter", it is a separate issue that this test is supposed to help isolate for future investigation.

Comment (toddbellMSFT) [reply source/test/TestApp/PlayFabGroupsTest.cpp:59]: We should be able to run these tests with local values, but attempting to do so ran into odd timeout issues that I wanted to investigate at a later time. It seems like a threading issue since debugging will pass the test, but running without stepping through times the test out. It looks very similar to an issue I was seeing with QoS, but it is out of scope to this particular PR.

**Review by MichaelMackendMS — APPROVED** [review]
thx for addressing the things!

File: source/test/TestApp/PlayFabGroupsTest.cpp (hunk 10 above)
Comment (pgilmorepf) [reply source/test/TestApp/PlayFabGroupsTest.cpp:49]: Actually, That is my point. Every other suite has login as a separate test & step. This is the only suite that would combine them, and I don't think it's a good idea. And, since this is a new suite, I don't think we should create it from scratch the wrong way, regardless.

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        AddTest("CreateGroupApi", &PlayFabGroupsTest::GroupsApiTest);
+    }
+
+    void PlayFabGroupsTest::ClassSetUp()
+    {
+        groupsTestSettings = std::make_shared<PlayFabApiSettings>();
+        groupsTestSettings->titleId = testTitleData.titleId;
+
+        clientApi = std::make_shared<PlayFabClientInstanceAPI>(groupsTestSettings);
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:57]: I could go either way but:
This constructor makes the clientApi utilize the staticPlayer, rather than a context defined in this test.
2 lines up, create an additional PlayFabAuthenticationContext, and pass it as a second parameter here.

Comment (pgilmorepf) [reply source/test/TestApp/PlayFabGroupsTest.cpp:57]: Additionally, the groupApi should be defined here, and not as a 1-off in the test above.

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        auto req = ClientModels::LoginWithCustomIDRequest();
+        req.CustomId = PlayFabSettings::buildIdentifier;
+
+        clientApi->LoginWithCustomID(req, Callback(&PlayFabGroupsTest::GroupsTestLoginCallback), Callback(&PlayFabGroupsTest::GroupsTestSharedFailureCallback), &testContext);
+    }
+
+    void PlayFabGroupsTest::GroupsTestLoginCallback(const ClientModels::LoginResult& result, void* customData)
+    {
+        auto groupApi = std::make_shared<PlayFabGroupsInstanceAPI>(result.authenticationContext);
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:24]: Move this into the ClassSetup, move the groupApi variable into the test class, and construct them both with shared settings and context.
I think that's why you were having such issues with the titleID and whatnot.
This one is being created with staticSettings, and clientApi is being created with a custom context, so they're not linked to the same information.

You're constructing the API instances completely differently, and that's the source of your data desync.
Your solution of result.authenticationContext HELPS, but it fails to share settings.
It's better to just create them together in ClassSetup.

File: source/test/TestApp/PlayFabGroupsTest.h [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+
+        void GroupsTestLoginCallback(const PlayFab::ClientModels::LoginResult& result, void* customData);
+        void GroupsTestSuccessCallback(const PlayFab::GroupsModels::CreateGroupResponse& response, void* customData);
+        void GroupsTestSharedFailureCallback(const PlayFab::PlayFabError& error, void* customData);
+
+    protected:
+        void AddTests() override;
+
+    public:
+        std::shared_ptr<PlayFab::PlayFabClientInstanceAPI> clientApi;
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.h:41]: add groupApi definition here.
```suggestion
        std::shared_ptr<PlayFab::PlayFabClientInstanceAPI> clientApi;
        std::shared_ptr<PlayFab::PlayFabGroupInstanceAPI> groupApi;
```

**Review by pgilmorepf — CHANGES_REQUESTED** [review]
I figured out why you were having data inconsistency problems with your titleID.
You're not using the same settings between your login and your create group.

File: source/test/TestApp/TestUtils.cpp [hunk 24]
```diff
@@ -1,10 +1,43 @@
...
 #include "TestAppPch.h"
-#include "TestUtils.h"
 
 namespace PlayFabUnit
 {
+    // source: https://stackoverflow.com/questions/24365331/how-can-i-generate-uuid-in-c-without-using-boost-library
+    // note this is not sufficiently random enough to be guid's, but random strings is the original need for our tests.
+    std::string GenerateRandomString() {
+        std::random_device              rd;
+        std::mt19937                    gen(rd());
```
Comment (toddbellMSFT) [inline source/test/TestApp/TestUtils.cpp:11]: _x64

File: source/test/TestApp/PlayFabGroupsTest.h [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+    protected:
+        std::shared_ptr<PlayFab::PlayFabClientInstanceAPI> clientApi;
+        std::shared_ptr<PlayFab::PlayFabGroupsInstanceAPI> groupsApi;
+
+        PlayFab::GroupsModels::EntityKey createdGroupKey;
+
+        void AddTests() override;
+
+    public:
+       void ClassSetUp() override;
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.h:46]: tabbing inconsistency. Press Ctrl+k+d to resolve all spacing issues with the whole file.

File: source/test/TestApp/PlayFabGroupsTest.h [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+            return std::bind(func, this, std::placeholders::_1, std::placeholders::_2);
+        }
+
+    protected:
+        std::shared_ptr<PlayFab::PlayFabClientInstanceAPI> clientApi;
+        std::shared_ptr<PlayFab::PlayFabGroupsInstanceAPI> groupsApi;
+
+        PlayFab::GroupsModels::EntityKey createdGroupKey;
+
+        void AddTests() override;
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.h:43]: *headscratch* Is AddTests a protected function? I thought it would be public...

File: source/test/TestApp/PlayFabGroupsTest.cpp [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+        testContext->Pass();
+    }
+
+    void PlayFabGroupsTest::GroupsRemoveTest(TestContext& testContext)
+    {
+        auto req = GroupsModels::DeleteGroupRequest();
+        req.CustomTags = std::map<std::string, std::string>();
+
+        // TODO Bug 29786037: this map is required to be filled to not get a 500 error with the CreateGroup api call
+        req.CustomTags.insert(std::pair<std::string, std::string>("One", "Two"));
```
Comment (pgilmorepf) [inline source/test/TestApp/PlayFabGroupsTest.cpp:53]: Wait, this is an issue with delete too?

**Review by pgilmorepf — APPROVED** [review]
Trivial notes. Maybe something actionable, especially the spacing issue, but otherwise ready to go.

**Review by Arnou2inf — APPROVED** [review]

**Comment by toddbellMSFT** [comment]
This was deemed out of scope. Closing.
